network/p2p/acp118/aggregator.go

```diff
@@ -0,0 +1,254 @@
+	}
+
+	requestBytes, err := proto.Marshal(request)
+	if err != nil {
+		return nil, 0, 0, fmt.Errorf("failed to marshal signature request: %w", err)
+	}
+
+	nodeIDsToValidator := make(map[ids.NodeID]indexedValidator)
+	// TODO expose concrete type to avoid type casting
+	bitSetSignature, ok := message.Signature.(*warp.BitSetSignature)
+	if !ok {
+		return nil, 0, 0, errors.New("invalid warp signature type")
+	}
+
+	var signerBitSet set.Bits
+	if bitSetSignature.Signers != nil {
+		signerBitSet = set.BitsFromBytes(bitSetSignature.Signers)
+	} else {
+		signerBitSet = set.NewBits()
+	}
+
+	sampleable := make([]ids.NodeID, 0, len(validators))
+	aggregatedStakeWeight := uint64(0)
+	totalStakeWeight := uint64(0)
+	for i, v := range validators {
+		totalStakeWeight += v.Weight
+
+		// Only try to aggregate signatures from validators that are not already in
+		// the signer bit set
+		if signerBitSet.Contains(i) {
+			aggregatedStakeWeight += v.Weight
+			continue
+		}
+
+		nodeIDsToValidator[v.NodeID] = indexedValidator{
+			Index:     i,
+			Validator: v,
+		}
+		sampleable = append(sampleable, v.NodeID)
+	}
+
+	signatures := make([]*bls.Signature, 0, len(sampleable)+1)
+	if bitSetSignature.Signature != [bls.SignatureLen]byte{} {
+		blsSignature, err := bls.SignatureFromBytes(bitSetSignature.Signature[:])
+		if err != nil {
+			return nil, 0, 0, fmt.Errorf("failed to parse bls signature: %w", err)
+		}
+		signatures = append(signatures, blsSignature)
+	}
+
+	results := make(chan result)
+	job := responseHandler{
+		message:            message,
+		nodeIDsToValidator: nodeIDsToValidator,
+		results:            results,
+	}
+
+	for _, nodeID := range sampleable {
+		// Avoid var shadowing in goroutine
+		nodeIDCopy := nodeID
+		go func() {
+			if err := s.client.AppRequest(ctx, set.Of(nodeIDCopy), requestBytes, job.HandleResponse); err != nil {
+				results <- result{Validator: nodeIDsToValidator[nodeIDCopy], Err: err}
+				return
+			}
+		}()
+	}
+
+	failedStakeWeight := uint64(0)
+	minThreshold := (totalStakeWeight * quorumNum) / quorumDen
+
+	for {
+		select {
+		case <-ctx.Done():
+			if len(signatures) == 0 {
+				return message, 0, totalStakeWeight, nil
+			}
+
+			// Try to return whatever progress we have if the context is cancelled
+			msg, err := newWarpMessage(message, signerBitSet, signatures)
+			if err != nil {
+				return nil, 0, 0, err
+			}
+
+			return msg, aggregatedStakeWeight, totalStakeWeight, nil
+		case result := <-results:
+			if result.Err != nil {
+				s.log.Debug(
+					"dropping response",
+					zap.Stringer("nodeID", result.Validator.NodeID),
+					zap.Uint64("weight", result.Validator.Weight),
+					zap.Error(err),
+				)
+
+				// Fast-fail if it's not possible to generate a signature that meets the
+				// minimum threshold
+				failedStakeWeight += result.Validator.Weight
+				if totalStakeWeight-failedStakeWeight < minThreshold {
+					return nil, 0, 0, ErrFailedAggregation
+				}
+				continue
```

Comment:
Does this work with hypersdk's expected usage here? I thought that `num/dem` were going to be the maximum weights it would wait for, but that the minimum would be lower than that (meaning that if we are passing in the max here, we could be terminating when we realize we can't get the maximum... But we actually could have gotten the number that hypersdk wanted).

Reply:
Maybe it makes more sense for us to change the behavior so that this api blocks until all responses come back, or we reach the provided `num/den` threshold instead.